raft: support lazy replication

pav-kv opened the pull request:
This PR adds support for `LazyReplication` mode in raft. In this mode, the leader does not send `MsgApp` messages eagerly to `StateReplicate` peers. Instead, the appends are requested explicitly by the application layer, via the `RawNode.SendMsgApp()` method.

The application:
- monitors the replication flow state to all followers, e.g. using the `RawNode.Status()` method,
- for peers who have outstanding replication work, decides when and how much worth of the log to send out,
- fetches the log slices from the `RawNode.LogSnapshot()`,
- sends the appends using the `RawNode.SendMsgApp()` method.

Part of #128779

pav-kv committed:
raft: support lazy replication

This commit adds support for LazyReplication mode in raft. In this mode,
the leader does not send MsgApp messages eagerly to StateReplicate
peers. Instead, the appends are requested explicitly by the application
layer, via the RawNode.SendMsgApp() method.

The application:
- monitors the replication flow state to all followers, e.g. using the
  RawNode.Status() method,
- for peers who have outstanding replication work, decides when and how
  much worth of the log to send out,
- fetches the log slice from the RawNode.LogSnapshot(),
- sends the appends using the RawNode.SendMsgApp() method.

Epic: none
Release note: none

pav-kv committed:
rafttest: test lazy replication

Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/131588)
<!-- Reviewable:end -->

pav-kv committed:
testdata: add basic replication test

This commits adds a prep test for the next commit, to demonstrate the
difference between the eager and lazy replication.

Epic: none
Release note: none

pav-kv committed:
raft,tracker: split entries vs probe sending logic

Epic: none
Release note: none

tbg reviewed (commented):
Looks good - appreciate the clean commits.

I'm mostly still unsure why it looks like we can send nonempty probes in lazy mode. Once that is resolved this PR is ready as far as I'm concerned.

pav-kv commented on pkg/raft/testdata/lazy_replication.txt:
```diff
@@ -0,0 +1,123 @@
+status 1
+----
+1: StateReplicate match=13 next=14 sentCommit=11 matchCommit=11
+2: StateReplicate match=13 next=14 sentCommit=13 matchCommit=13
```
Yes, I tripped on this too. We discussed this effect with @iskettaneh a bunch while he was working on #129943, and agreed that the leader's commit indices in `Progress` struct is ok to be stale. The leader does not use its own indices for replication (there are no explicit self-directed `MsgApp`s on the leader that would depend on them). Some background:

`matchCommit` is a commit index known to be durably persisted on this peer (we use this fact to quiesce `MsgApp` pings when the peer is fully up-to-date). On the leader, we don't know if the commit index was synced, and don't enforce it. If we enforced it (for consistency, in order to make this `matchCommit` converge), it could cause unnecessary syncs in some cases.

We could update `sentCommit` to be equal to `raftLog.commit` at least. But since `matchCommit` isn't converging, we would similarly be "confused" when looking at this line. The best we would be able to achieve here is that `sentCommit == raftLog.commit`, and `matchCommit` is lagging by one `MsgStorageAppend` roundtrip.

pav-kv commented on pkg/raft/testdata/lazy_replication.txt:
```diff
@@ -0,0 +1,123 @@
+----
+ok
+
+add-nodes 3 voters=(1,2,3) index=10 lazy-replication=true
```
Fixed.

pav-kv commented on pkg/raft/tracker/progress.go:
```diff
@@ -304,47 +304,52 @@ func (pr *Progress) IsPaused() bool {
+		return !pr.MsgAppProbesPaused && pr.CanSendEntries(last)
+	case StateReplicate:
+		return !lazyReplication && pr.CanSendEntries(last)
+	case StateSnapshot:
```
Filed #131681 to this extent.

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 7 files at r5, 2 of 2 files at r6, 2 of 2 files at r7, 4 of 5 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131588#-:-O89FtkB9qW7j67l26fE:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @iskettaneh and @tbg)



<!-- Sent from Reviewable.io -->

tbg commented on pkg/raft/tracker/progress.go:
```diff
@@ -304,47 +304,51 @@ func (pr *Progress) IsPaused() bool {
+	switch pr.State {
+	case StateProbe:
+		return !pr.MsgAppProbesPaused && pr.CanSendEntries(last)
+	case StateReplicate:
```
Thank you!

tbg reviewed (approved):
Thank you for the explainers!

pav-kv commented:
bors r=tbg,sumeerbhola
